File: Cpp/525.cpp
#include<iostream>
#include<vector>
#include<unordered_map>

using namespace std;

class Solution{
public:
	int findMaxLength(vector<int>& nums){
		unordered_map<int,int> m;
		int sum=0;
		int maxLen=0;
		m[0]=-1;
		for(int i=0;i<nums.size();++i){
			sum+=nums[i]==0?-1:1;
			auto it=m.find(sum);
			if(it!=m.end()){
				maxLen=max(maxLen,i-m[sum]);
			}
			else
				m[sum]=i;
		}
		return maxLen;
	}
};

int main(){
	vector<int> nums={0,0,0,0,1,1};
	Solution s;
	cout<<s.findMaxLength(nums)<<endl;

	return 0;

}

File: Python/525.py
class Solution(object):
    def findMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        sum = 0
        m = {0:-1}
        maxLen = 0
        for i, n in enumerate(nums):
            if n == 0:
                sum -=1
            else:
                sum +=1
            if sum not in m:
                m[sum] = i
            else:
                maxLen = max(maxLen, i-m[sum])
        return maxLen
